packages/flutter_tools/lib/src/build_system/targets.dart

```diff
@@ -0,0 +1,329 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../asset.dart';
+import '../base/file_system.dart';
+import '../base/io.dart';
+import '../base/process_manager.dart';
+import '../build_info.dart';
+import '../compile.dart';
+import '../devfs.dart';
+import '../project.dart';
+
+import 'build_system.dart';
+
+/// Build definitions.
+
+/// List all asset files in a project by parsing the asset manfiest.
+List<File> listAssets(Environment environment) {
+  final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
+  assetBundle.build(
+    manifestPath: environment.projectDir.childFile('pubspec.yaml').path,
+    packagesPath: environment.projectDir.childFile('.packages').path,
+  );
+  final List<File> results = <File>[];
+  final Iterable<DevFSFileContent> files = assetBundle.entries.values.whereType<DevFSFileContent>();
+  for (DevFSFileContent devFsContent in files) {
+    results.add(fs.file(devFsContent.file.path));
+  }
+  return results;
+}
+
+/// List all output files in a project by parsing the asset manfiest and
+/// replacing the path with the output directory.
+List<File> listOutputAssets(Environment environment) {
+  final List<File> inputs = listAssets(environment);
+  final List<File> results = <File>[];
+  for (File input in inputs) {
+    final String assetName = fs.path.join(
+      environment.buildDir.path,
+      'flutter_assets',
+      fs.path.basename(input.path),
+    );
+    results.add(fs.file(assetName));
+  }
+  return results;
+}
+
+/// Assemble the assets used in the application into a build directory.
+const Target copyAssets = Target(
+  name: 'copy_assets',
+  inputs: <Object>[
+    '{PROJECT_DIR}/pubspec.yaml',
+    listAssets,
+  ],
+  outputs: <Object>[
+    '{BUILD_DIR}/flutter_assets/AssetManifest.json',
+    '{BUILD_DIR}/flutter_assets/FontManifest.json',
+    '{BUILD_DIR}/flutter_assets/LICENSE',
+    listOutputAssets, // <- everything in this subdirectory.
+  ],
+  dependencies: <Target>[],
+  invocation: copyAssetsInvocation,
+);
+
+/// Copies the asset files from the [copyAssets] rule into place.
+Future<void> copyAssetsInvocation(List<FileSystemEntity> inputs, Environment environment) async {
+  final Directory output = environment.buildDir.childDirectory('flutter_assets');
+  if (!output.existsSync()) {
+    output.createSync(recursive: true);
+  }
+  final AssetBundle assetBundle = AssetBundleFactory.instance.createBundle();
+  await assetBundle.build(
+    manifestPath: environment.projectDir.childFile('pubspec.yaml').path,
+    packagesPath: environment.projectDir.childFile('.packages').path,
+  );
+  await Future.wait<void>(
+    assetBundle.entries.entries.map((MapEntry<String, DevFSContent> entry) async {
+      final File file = fs.file(fs.path.join(output.path, entry.key));
```

Comment:
macOS and (in the future Fuchsia) systems have limited numbers of file descriptors.